### ISSUE TITLE
Can't input Tilde (~) character with pt-br keyboard

### ISSUE THREAD

**Issue by HenriAugusto** [pure-data/pure-data#732]
I'm running pd-0.50-0test1 and i can't type tildes anymore while using the PT-BR keyboard.

I can only type them if i change my system's keyboard setting to "EN".

Here are images of both layouts, just in case somebody wants to see them
[US keyboard](https://en.wikipedia.org/wiki/British_and_American_keyboards#/media/File:KB_United_States-NoAltGr.svg) | [PT-BR keyboard](https://en.wikipedia.org/wiki/Portuguese_keyboard_layout#/media/File:KB_Portuguese_Brazil.svg)

In english you just press **shift + '** and the tilde is output directly.
In portuguese we have chars like "ã" and "õ"so here is how it works: first you type the "~" key (and nothing happens) and then _a_ or _o_ so it ouputs _ã_ or _õ_. Or if you want to type the "~" alone (as we do in PD) you just press the "~" key twice. 

When pressing it twice, as reported by the [key] obj:

In pd 0.49 the 1st time press the "~" key i get keycode 0 and then the 2nd time i get 126. An "~" is output

In pd 0.50 the 1st time press the "~" key i get keycode 0 and then the 2nd time i get 0 again. Nothing is output.

**Comment by danomatika** [comment]
Does it work if you type into a GUI text field, ie. the find window? If it doesn't;t, then the problem is likely with Tk.

**Comment by sevdh** [comment]
What happens if you press ~ followed by a space?

**Comment by HenriAugusto** [comment]
> Does it work if you type into a GUI text field, ie. the find window? If it doesn't, then the problem is likely with Tk.

Yes. It outputs two tildes `~~` which is the same behavior throghout windows. I'm not sure how in PD it would only output one tilde `~`, which was way better.

> What happens if you press ~ followed by a space?

In an object (pd-0.50): it outputs a space
In an object (pd-0.49): outputs only a single tilde.
Throghout windows: outputs only a single tilde.

**Comment by umlaeute** [comment]
i can reproduce the issue on Windows 10.

Pressing the <kbd>~</kbd> key (just left of <kbd>Return</kbd> ) twice, the GUI sends:

~~~
.x24a680 key 1 Multi_key 0;
.x24a680 key 0 Multi_key 0;

.x24a680 key 1 Multi_key 0;
.x24a680 key 0 Multi_key 0;
~~~

(the empty indicates the pause between the two keypresses)

OTOH, when i run Pd-0.50test1 under wine on Debian (for whatever reasons only the 32bit version starts, up), I see this instead when i double-push the same key, and I correctly get a `~`:

~~~
.x24a680 key 0 Multi_key 0;

.x24a680 key 1 126 0;
.x24a680 key 0 Multi_key 0;
~~~

So:
- on W10
  - the `Multi_key` sends both key-press and key-release events
  - but tilde `~` cannot be typed
  - other deadkeys work fine (e.g. `~` + `n` gives me `ñ`)
- on Linux/Wine:
  - the `Multi_key` sends only key-release events
  - but on re-pressing the key, a compound character `~` is typed
  - other deadkeys work fine (e.g. `~` + `n` gives me `ñ`)

**Comment by umlaeute** [comment]
sure enough the problem comes from my fixes around b713e20796d993622f4b66a17e9a0b610b6065a4.

That Pd<=0.49 would output a single tilde was an accident: it would output the first character of a "multicharacter event". in your case (reducing `~~` to `~`) it was nice, but in general it was rather annoying (e.g. reducing `space` (indicating the space-bar) to `s`).
so i think that *this* should be gone for good.

but of course, we must find a way to allow portuguese people to type `~`.

**Comment by porres** [comment]
well, this is kind of a show stopper now that pd 0.50 is out. I already see people in the brazilian Pd community complaining about this and saying they'll stick to Pd 0.49 for now. Could we start a new 0.50.1 branch and include this issue as a bug fix priority?

**Comment by umlaeute** [comment]
Pd development has no notion of "high priority bug fix branch".
just create an ordinary PR that *fixes* the problem. it will then be included (or not, depending on the existence of other solutions for the problem at hand) whenever the next bugfix release gets out.

**Comment by porres** [comment]
What I mean is a milestone, such as https://github.com/pure-data/pure-data/milestone/6 - I consider this a priority list, and also, usually you people setup a branch for 'bugfixes' and include many things there. So this is what I meant. After all this is in fact a bug that was introduced in 0.50, so it's one thing that could be fixed in a possible 0.50.1 bugfix release.

**Comment by umlaeute** [comment]
we have the `updates/0.50` branch, which is still up and a running (it didn't get fully merged into `0.50`). so what is wrong with that?

the *milestones* really are just a feature of github, that we are sometimes using "for the fun of it" (at least on my side, i cannot speak for @danomatika, and i don't think anybody else ever used it), because github offers it, and because it is sometimes a nice way to group things by "topic" rather than more format categories ("bug" , "enhancement").
i don't think that assigning anything to a "milestone" will give it any priority.

> After all this is in fact a bug that was introduced in 0.50, so it's one thing that could be fixed in a possible 0.50.1 bugfix release.

no one is disputing that. why are you even mentioning it?

**Comment by porres** [comment]
Well, I think it's a nice feature of github, and I just thought they were used more "seriously". I just emphasized that it would perfectly belong to such a milestone, not for a dispute, just to corroborate my point that it would fit there. But I see now your point that they're not that crucial for the development, so it's fine, I don't care :) 

> just create an ordinary PR that *fixes* the problem.

yeah, I'd totally do that and be over with, but I don't think I can for this one, so I just hoped that it could sit in a more visible spot so it's on the radar of the proper Pd gurus to attack it.

cheers

**Comment by HenriAugusto** [comment]
Since this is a very specific problem maybe it's the case for an _ad hoc_ solution?

```TCL
    if { [string length $iso] == 1 } {
        scan $iso %c key
    } elseif { $iso eq "~~"} {
        #tilde = 126
        set key 126 
    }
```

It seems on windows  the "~" will always arrive with as "Multi_key" so we need to read the iso. And the iso will only be a "~~". At least i couldn't get it to arrive as a "~" with a single press while messing with the bindings (i know from experience some things might behave differently on <Key> and <KeyRelease> events, for example)

**Comment by umlaeute** [comment]
> Since this is a very specific problem maybe it's the case for an _ad
>hoc_ solution?


i'm afraid the problem is less specific (dead key handling with pressing the same (dead) key twice.)
it could happen on any keyboard layout, and not just with the `~`

**Comment by umlaeute** [comment]
could you try whether the [deadkeys](https://github.com/pure-data/pure-data/tree/deadkeys) branch fixes the problem?

i would also welcome tests on other platforms (apple!) on keyboard layouts with and without dead-keys (english, *french*, ...)

**Comment by phil123456** [comment]
lol, I was installing PD on another PC, thinking that I'd be better off installing a version I knew
fucking regressions

**Comment by umlaeute** [comment]
thanks @phil123456 for your valuable feedback.
you might have noticed that the problem has been fixed meanwhile
